Tasks/Maven/mavencodeanalysis.ts

```diff
@@ -0,0 +1,231 @@
+
+        // Discover output files, summarise per-module results
+        modules.forEach((module:ma.ModuleAnalysis) => {
+            // PMD
+            if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
+                try {
+                    var pmdResults:ar.AnalysisResult = pmd.collectPmdOutput(module.rootDirectory);
+                    if (pmdResults) {
+                        module.analysisResults[pmdResults.toolName] = pmdResults;
+                    }
+                } catch(e) {
+                    tl.error(e.message);
+                    tl.exit(1);
+                }
+            }
+
+        });
+
+
+        // Gather analysis results by tool for summaries, stage and upload build artifacts
+        var buildSummaryLines:string[] = [];
+        cleanDirectory(stagingDir);
+
+        enabledCodeAnalysisTools.forEach((toolName:string) => {
+            var toolAnalysisResults:ar.AnalysisResult[] = [];
+            modules.forEach((module:ma.ModuleAnalysis) => {
+                var moduleAnalysisResult:ar.AnalysisResult = module.analysisResults[toolName];
+                if (moduleAnalysisResult) {
+                    toolAnalysisResults.push(moduleAnalysisResult);
+                }
+
+                uploadBuildArtifacts(toolName, module);
+            });
+
+            // After looping through all modules, summarise tool output results
+            try {
+                var summaryLine:string = createSummaryLine(toolName, toolAnalysisResults);
+            } catch (error) {
+                tl.error(error.message);
+            }
+            buildSummaryLines.push(summaryLine);
+        });
+
+        // Save and upload build summary
+        // Looks like: "PMD found 13 violations in 4 files.  \r\n
+        // FindBugs found 10 violations in 8 files."
+        var buildSummaryContents:string = buildSummaryLines.join("  \r\n"); // Double space is end of line in markdown
+        var buildSummaryFilePath:string = path.join(stagingDir, 'CodeAnalysisBuildSummary.md');
+        fs.writeFileSync(buildSummaryFilePath, buildSummaryContents);
+        tl.debug('Uploading build summary from ' + buildSummaryFilePath);
+
+        tl.command('task.addattachment', {
+            'type': 'Distributedtask.Core.Summary',
+            'name': "Code Analysis Report"
+        }, buildSummaryFilePath);
+    }
+}
+
+// Returns the names of any enabled code analysis tools, or empty array if none.
+function getEnabledCodeAnalysisTools():string[] {
+    var result:string[] = [];
+
+    if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
+        console.log('PMD analysis is enabled');
+        result.push(pmd.toolName);
+    }
+
+    return result;
+}
+
+// Returns the full path of the staging directory for a given tool.
+function getStagingDirectory(toolName:string):string {
+    return path.join(stagingDir, toolName.toLowerCase());
+}
+
+function cleanDirectory(targetDirectory:string):boolean {
+    tl.rmRF(targetDirectory);
```

Comment:
I have a suspicion it will look something like the Node ENOENT error. That said, no attempt to do anything better is seen as an example in any of the other TS tasks, and no errors have ever been observed in testing.